telcell/utils/transform.py

```diff
@@ -0,0 +1,222 @@
+    :return: A filtered list with all paired measurements.
+    """
+    return [x for x in paired_measurements
+            if x.time_difference <= max_delay]
+
+
+def sort_pairs_based_on_rarest_location(
+        switches: List[MeasurementPair],
+        history_track_b: Track,
+        round_lon_lats: bool,
+        max_delay: int = None
+) -> List[Tuple[int, MeasurementPair]]:
+    """
+    Pairs are first filtered on allowed time interval of the two registrations
+    of a single pair. Then, sort pairs based on the rarest location of the
+    track history first and secondly by time difference of the pair.
+
+    :param switches: A list with all paired measurements to consider.
+    :param history_track_b: the history of track_b to find the rarity of locations.
+    :param round_lon_lats: boolean indicating whether to round the lon/lats
+            to two decimals.
+    :param max_delay: maximum allowed time difference (seconds) in a pair.
+                      Default: no max_delay, show all possible pairs.
+    :return: The location counts and measurement pairs that are sorted on the
+            rarest location based on the history and time difference. The
+            location count is the number of occurrences of the coordinates from
+            measurement_b in the track history that is provided.
+
+    TODO There is a problem with testdata, because those are almost continuous
+    lat/lon data, making rarity of locations not as straightforward.
+    Pseudo-solution for now: round lon/lats to two decimals and determine
+    rarity of those.
+    This should not be used if locations are actual cell-ids
+    """
+
+    def location_key(measurement):
+        if round_lon_lats:
+            return f'{measurement.lon:.2f}_{measurement.lat:.2f}'
+        else:
+            return f'{measurement.lon}_{measurement.lat}'
+
+    def sort_key(element):
+        rarity, pair = element
+        return rarity, pair.time_difference
+
+    location_counts = Counter(
+        location_key(m) for m in history_track_b.measurements)
+
+    if max_delay:
+        switches = filter_delay(switches, timedelta(seconds=max_delay))
+
+    sorted_pairs = sorted(
+        ((location_counts.get(location_key(pair.measurement_b), 0), pair) for
+         pair in switches), key=sort_key)
+
+    return sorted_pairs
+
+
+def select_colocated_pairs(tracks: List[Track],
+                           max_delay: timedelta = timedelta(seconds=120)) \
+        -> List[MeasurementPair]:
+    """
+    For a list of tracks, find pairs of measurements that are colocated, i.e.
+    that do not share the same track name, but do share the owner. Also filter
+    the pairs based on a maximum time delay.
+
+    :param tracks: the tracks to find pairs of.
+    :param max_delay: the maximum amount of delay that is allowed.
+    :return: A filtered list with all colocated paired measurements.
+    """
+    tracks_per_owner = defaultdict(list)
+    for track in tracks:
```

Comment:
We hebben er bewust voor gekozen om `is_colocated` expliciet als functie te definieren. Deze zou dan ook hier gebruikt moeten worden maar wordt nu omzeild.

Reply:
opgelost!